docs/proposals/network-policy.md

```diff
@@ -0,0 +1,300 @@
+one or more NetworkPolicy objects should allow any incoming connections that match any
+ingress rule on those NetworkPolicy objects, per the network plugin’s capabilities.
+
+If `ingress-isolation` is not specified on a namespace, then all traffic is allowed to pods in that namespace.
+
+```go
+type NetworkPolicy struct {
+  TypeMeta
+  ObjectMeta
+
+  // Specification of the desired behavior for this NetworkPolicy.
+  Spec NetworkPolicySpec 
+}
+
+type NetworkPolicySpec struct {
+  // Selects the pods to which this NetworkPolicy object
+  // applies.  The array of NetworkPolicyIngressRules below 
+  // is applied to any pods selected by this field.
+  // Multiple NetworkPolicy objects can select the same set of pods.  In this case,
+  // the NetworkPolicyRules for each are combined additively.
+  PodSelector *unversioned.LabelSelector `json:"podSelector,omitempty"`
+
+  // List of ingress rules to be applied to the selected pods.
+  // Traffic is allowed to a pod if Namespace.NetworkPolicy.Ingress.Isolation is undefined, 
+  // OR if the traffic source is the pod's host (for kubelet health checks), 
+  // OR if Namespace.NetworkPolicy.Ingress.Isolation=DefaultDeny and the traffic matches at least 
+  // one NetworkPolicyIngressRule across all of the NetworkPolicy 
+  // objects whose podSelector matches the pod.  If this field is 
+  // empty, this NetworkPolicy has no effect on selected pods.
+  Ingress []NetworkPolicyIngressRule `json:"ingress,omitempty"`
+}
+
+type NetworkPolicyIngressRule struct {
+  // List of ports which should be made accessible on the pods selected by PodSelector.  
+  // Each item in this list is combined using a logical OR.  If this field
+  // is empty, then traffic should not be restricted based on port.  If this 
+  // field is not empty, traffic must match at least one NetworkPolicyPort in the list
+  // or else it will be dropped.
+  Ports []NetworkPolicyPort `json:"ports,omitempty"`
+
+  // List of sources which should be able to access the pods selected by PodSelector.
+  // Items in this list are combined using a logical OR operation.
+  // If this field is empty, then traffic should not be restricted based on 
+  // source.  If this field is not empty, traffic must match at least one 
+  // NetworkPolicySource in the list or else it will be dropped.
+  From []NetworkPolicySource `json:"from,omitempty"`
+}
+
+type NetworkPolicyPort struct {
+  // The protocol (TCP or UDP) which traffic must match.
+  // If not defined, this field defaults to TCP. 
+  Protocol api.Protocol `json:"protocol"`
+
+  // If specified, the port on the given protocol.  This can 
+  // either be a numerical or named port.  If not defined,
+  // this NetworkPolicyPort does not restrict based on port number.
+  // If defined, only traffic on the specified protocol AND port
+  // will be matched by this NetworkPolicyPort.
+  Port *intstr.IntOrString `json:"port,omitempty"`
+}
+
+type NetworkPolicySource struct {
```

Comment:
Another unresolved discussion was cut off by the latest edit.  This is the one about how a `&NetworkPolicySource{}` (that is, one with both fields omitted) matches all sources (according to the definition here) --- and this is a better way to say "match all sources" than omitting the `From` field of a `NetworkPolicyIngressRule`.  I noted the mental irregularity of making missing `From` mean match all sources while a non-missing `From` matches less.  I recommended allowing only one way to say "match all sources", namely a `NetworkPolicySource` with both fields absent.  Casey agreed that this seems preferable but wanted additional review from Tim and other API guardians.  Dan replied with something I do not understand about potential trouble when egress is introduced; I see no way that this could lead to trouble when egress is introduced.

BTW, we could do the same thing on the `Ports` side.  Just allow the `Protocol` field of a `NetworkPolicyPort` to be omitted with the meaning that matching is not restricted by protocol.  Then a `NetworkPolicyPort` with both fields omitted will match all port/protocol, and we do not need the irregularity of a missing `NetworkPolicyIngressRule.Ports` matching more than a non-missing `NetworkPolicyIngressRule.Ports`.